File: apps/web/src/shared/api/agentRuns.ts
/**
 * API client for Agent Runs observability
 */
import { apiRequest } from './http';
import type { RunTrace } from '@/domains/runtimeTrace/types';

export interface AgentRunStep {
  id: string;
  step_number: number;
  step_type:
    | 'user_request'
    | 'routing'
    | 'llm_request'
    | 'llm_response'
    | 'llm_call'
    | 'tool_call'
    | 'tool_result'
    | 'operation_call'
    | 'operation_result'
    | 'protocol_retry'
    | 'budget_snapshot'
    | 'final'
    | 'final_response'
    | 'error'
    | 'run_start'
    | 'run_end'
    | 'orchestrator_start'
    | 'orchestrator_end'
    | 'agent_start'
    | 'agent_end'
    | 'synthesis_start'
    | 'synthesis_end'
    | string;
  data: Record<string, unknown>;
  tokens_in?: number;
  tokens_out?: number;
  duration_ms?: number;
  error?: string;
  created_at: string;
}

export interface AgentRunContextSnapshot {
  agent_slug?: string;
  agent_version_id?: string;
  prompt_hash?: string;
  model?: string;
  execution_mode?: string;
  policy?: {
    max_steps?: number;
    max_tool_calls?: number;
    max_wall_time_ms?: number;
    tool_timeout_ms?: number;
    streaming_enabled?: boolean;
  };
  tools?: Array<{ slug: string; instance_id?: string; has_credentials?: boolean }>;
  available_operations?: Array<string | {
    operation_slug?: string;
    operation?: string;
    tool?: string;
    name?: string;
    canonical_name?: string;
    scope_kind?: 'system' | 'collection';
    title?: string;
    description?: string;
    result_kind?: string;
    collection_slug?: string;
    collection_type?: string;
    collection_purpose?: string;
    collection_readiness?: string;
    schema_freshness?: string;
    provider_kind?: string;
    input_schema_summary?: string[];
    side_effects?: boolean;
    risk_level?: 'safe' | 'write' | 'destructive';
  }>;
  available_collections?: Array<{
    collection_slug: string;
    collection_type?: string;
    title?: string;
    purpose?: string;
    data_description?: string;
    readiness_status?: string;
    schema_freshness?: string;
    missing_requirements?: string[];
    available_operation_slugs?: string[];
  }>;
  routing_duration_ms?: number;
  routing_reasons?: string[];
  request_text?: string;
}

export interface AgentRun {
  id: string;
  chat_id?: string;
  message_id?: string;
  user_id?: string;
  tenant_id: string;
  agent_slug: string;
  logging_level?: 'none' | 'errors' | 'brief' | 'full';
  status: 'running' | 'completed' | 'failed';
  total_steps: number;
  total_tool_calls: number;
  total_llm_calls?: number;
  tokens_in?: number;
  tokens_out?: number;
  duration_ms?: number;
  error?: string;
  started_at: string;
  finished_at?: string;
  context_snapshot?: AgentRunContextSnapshot;
}

export interface AgentRunDetail extends AgentRun {
  steps: AgentRunStep[];
  trace?: RunTrace;
}

export interface AgentRunListResponse {
  items: AgentRun[];
  total: number;
  page: number;
  page_size: number;
}

export interface AgentRunFilter {
  tenant_id?: string;
  user_id?: string;
  chat_id?: string;
  agent_slug?: string;
  status?: string;
  from_date?: string;
  to_date?: string;
  page?: number;
  page_size?: number;
}

export interface AgentRunStats {
  total_runs: number;
  by_status: Record<string, number>;
  by_agent: Record<string, number>;
  avg_duration_ms?: number;
}

export const agentRunsApi = {
  async list(filters: AgentRunFilter = {}): Promise<AgentRunListResponse> {
    const params = new URLSearchParams();
    if (filters.tenant_id) params.set('tenant_id', filters.tenant_id);
    if (filters.user_id) params.set('user_id', filters.user_id);
    if (filters.chat_id) params.set('chat_id', filters.chat_id);
    if (filters.agent_slug) params.set('agent_slug', filters.agent_slug);
    if (filters.status) params.set('status', filters.status);
    if (filters.from_date) params.set('from_date', filters.from_date);
    if (filters.to_date) params.set('to_date', filters.to_date);
    if (filters.page) params.set('page', String(filters.page));
    if (filters.page_size) params.set('page_size', String(filters.page_size));
    
    const query = params.toString();
    const url = `/admin/agent-runs${query ? `?${query}` : ''}`;
    return apiRequest<AgentRunListResponse>(url);
  },

  async get(runId: string): Promise<AgentRunDetail> {
    return apiRequest<AgentRunDetail>(`/admin/agent-runs/${runId}`);
  },

  async delete(runId: string): Promise<{ status: string; run_id: string }> {
    return apiRequest(`/admin/agent-runs/${runId}`, { method: 'DELETE' });
  },

  async deleteOld(beforeDate: string, tenantId?: string): Promise<{ status: string; deleted_count: number }> {
    const params = new URLSearchParams({ before_date: beforeDate });
    if (tenantId) params.set('tenant_id', tenantId);
    return apiRequest(`/admin/agent-runs?${params.toString()}`, { method: 'DELETE' });
  },

  async getStats(filters: { tenant_id?: string; from_date?: string; to_date?: string } = {}): Promise<AgentRunStats> {
    const params = new URLSearchParams();
    if (filters.tenant_id) params.set('tenant_id', filters.tenant_id);
    if (filters.from_date) params.set('from_date', filters.from_date);
    if (filters.to_date) params.set('to_date', filters.to_date);
    
    const query = params.toString();
    return apiRequest<AgentRunStats>(`/admin/agent-runs/stats/summary${query ? `?${query}` : ''}`);
  },
};

File: apps/api/src/app/schemas/agent_runs.py
from typing import Any, Dict, List, Optional
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field


class AgentRunStepResponse(BaseModel):
    """Single step within an agent run"""
    id: UUID
    step_number: int
    step_type: str = Field(
        ...,
        description="user_request, routing, llm_request, llm_response, "
                    "tool_call, tool_result, final_response, error"
    )
    data: Dict[str, Any] = Field(default={})
    tokens_in: Optional[int] = None
    tokens_out: Optional[int] = None
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class AgentRunResponse(BaseModel):
    """Agent run summary"""
    id: UUID
    chat_id: Optional[UUID] = None
    message_id: Optional[UUID] = None
    user_id: Optional[UUID] = None
    tenant_id: UUID
    agent_slug: str
    logging_level: str = Field(default="brief", description="none, brief, full")
    status: str = Field(..., description="running, completed, failed")
    context_snapshot: Optional[Dict[str, Any]] = None
    total_steps: int = 0
    total_tool_calls: int = 0
    total_llm_calls: int = 0
    tokens_in: Optional[int] = None
    tokens_out: Optional[int] = None
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    started_at: datetime
    finished_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AgentRunDetailResponse(AgentRunResponse):
    """Agent run with all steps"""
    steps: List[AgentRunStepResponse] = []


class AgentRunListResponse(BaseModel):
    """Paginated list of agent runs"""
    items: List[AgentRunResponse]
    total: int
    page: int
    page_size: int


class AgentRunTracePackResponse(BaseModel):
    """Deterministic trace pack for replay/debug."""
    run_id: UUID
    agent_slug: str
    status: str
    logging_level: str
    context_snapshot: Optional[Dict[str, Any]] = None
    operations: List[str] = []
    prompt_surfaces: List[Dict[str, Any]] = []
    tool_io: List[Dict[str, Any]] = []
    errors: List[Dict[str, Any]] = []
    timeline: List[Dict[str, Any]] = []
    total_steps: int = 0


class CapabilityGraphNodeResponse(BaseModel):
    id: str
    type: str
    label: str
    meta: Dict[str, Any] = {}


class CapabilityGraphEdgeResponse(BaseModel):
    id: str
    from_: str = Field(alias="from")
    to: str
    type: str
    meta: Dict[str, Any] = {}

    class Config:
        populate_by_name = True


class RuntimeCapabilityGraphResponse(BaseModel):
    tenant_id: UUID
    user_id: UUID
    agent_slug: Optional[str] = None
    nodes: List[CapabilityGraphNodeResponse] = []
    edges: List[CapabilityGraphEdgeResponse] = []
    stats: Dict[str, int] = {}
    missing: Dict[str, List[str]] = {}


class RuntimeHitlConditionResponse(BaseModel):
    condition_id: str
    decision: str
    when: str
    reason: str


class RuntimeOperationPolicyResponse(BaseModel):
    operation_slug: str
    operation: str
    name: str
    side_effects: str
    risk_level: str
    requires_confirmation_semantic: bool
    effective_decision: str
    reasons: List[str] = []


class RuntimeHitlPolicyContractResponse(BaseModel):
    tenant_id: UUID
    user_id: UUID
    global_: Dict[str, Any] = Field(alias="global")
    conditions: List[RuntimeHitlConditionResponse] = []
    operation_policies: List[RuntimeOperationPolicyResponse] = []
    resume_contract: Dict[str, Any] = {}

    class Config:
        populate_by_name = True


class AgentRunFilter(BaseModel):
    """Filter options for agent runs"""
    tenant_id: Optional[UUID] = None
    user_id: Optional[UUID] = None
    chat_id: Optional[UUID] = None
    agent_slug: Optional[str] = None
    status: Optional[str] = None
    from_date: Optional[datetime] = None
    to_date: Optional[datetime] = None
